### ISSUE TITLE
Test bindings using both Dart-only and Dart+C modes

### ISSUE THREAD

**Issue by HosseinYousefi** [dart-lang/jnigen#183]
Currently some tests are happening in Dart-only mode and some other are happening in Dart+C. We're only running `dart analyze` on the other mode.

**Comment by mahesh-hegde** [comment]
When implementing #242, it's desirable to test both Dart+C and Dart-only bindings.

So what implementation strategy should I go with?

* Check in pure-dart and c-based bindings for all tests.

* Only check-in few bindings, generate the others on-the-fly.

**Comment by HosseinYousefi** [comment]
> When implementing #242, it's desirable to test both Dart+C and Dart-only bindings.
> 
> So what implementation strategy should I go with?
> 
> * Check in pure-dart and c-based bindings for all tests.
> * Only check-in few bindings, generate the others on-the-fly.

Let's focus on "runtime" tests. I think we don't want to duplicate the tests, so we would want to generate the bindings on the fly and test them using the same file. We might even generate the pure-dart and c-based ones in the same path so we don't even have to change the import. Another test could generate the bindings, analyze them, test them with the golden tests (if exist), and run the runtime tests for the generated bindings.

**Comment by HosseinYousefi** [comment]
> We don't want to modify any file that's checked-in to git.

Why not? We could remove the C-based bindings committed by default and change it to pure-Dart ones and rerun the runtime tests.

---

[This](https://github.com/dart-lang/language/issues/2235) would solve some of our intellisense issues.

**Comment by mahesh-hegde** [comment]
> Why not? We could remove the C-based bindings committed by default and change it to pure-Dart ones and rerun the runtime tests.

Because test would alter repository state, which may affect later tests, or cause the altered file to be committed accidentally.

The generated bindings which are checked in must be a "pure function" of config and source which are also checked in.

**Comment by mahesh-hegde** [comment]
How about this:

* Check in both type of bindings - this will allow us to see changes in both when doing code generator changes.

* Based on existing tests for one type (`c_based`), generate other type of bindings at runtime in files named `_replicated_test.dart`.
    * gitignore the pattern `*_replicated_test.dart`

* We already depend on summarizer being built before tests run, so why dont we wrap these into `tool/run_tests.dart` and error on `dart test` being run directly?
```dart
// dart run jnigen:setup
// generateReplicatedTests()
// dart test, stream results to stdout
// delete replicated bindings
```

**Comment by HosseinYousefi** [comment]
> * Check in both type of bindings - this will allow us to see changes in both when doing code generator changes.

That's good.

We can have some `pre_test.dart` that simply replicates the given runtime tests, and changes the import path. We can run this in the CI before running `dart test`. It's basically like having a copy of the test.

**Comment by HosseinYousefi** [comment]
> I'd suggest we just wrap all these into single script as above described, and either error on directly invoking dart test, telling to run tool/run_tests.dart instead.

Any reason to combine the two?

My reasoning for having a separate steps: `dart test` still will run fine, it still tests one of the two bindings, and all the VSCode test plugin integrations still work nicely with the test. The script is there to create the "missing" test from the "existing" one. The coverage gets calculated correctly in the CI.

**Comment by mahesh-hegde** [comment]
I actually ended up following the way you suggested, albeit with a few changes.

When `dart test` is run without pre-setup, no runtime tests are run. This is because we can have at max 1 JVM, and it needs to have all classpaths. (We can't add classPath in runtime). So `generated_runtime_test.dart` takes care of calling all tests defined in different files.

This is kind of similar to how we separated `runTests` in JNI.

This means VS code UI for running individual tests doesn't work very nicely, however. (I always use the command line, so never thought about it).
